File: src/main/java/org/openbase/planetsudo/level/levelobjects/Mothership.java
package org.openbase.planetsudo.level.levelobjects;

/*-
 * #%L
 * PlanetSudo GameEngine
 * %%
 * Copyright (C) 2009 - 2020 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.beans.PropertyChangeEvent;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import javax.swing.Timer;
import org.slf4j.Logger;
import org.openbase.planetsudo.game.GameManager;
import org.openbase.planetsudo.game.GameSound;
import org.openbase.planetsudo.game.Team;
import org.openbase.planetsudo.level.AbstractLevel;
import org.openbase.planetsudo.main.GUIController;
import java.util.ArrayList;
import org.openbase.jul.exception.CouldNotPerformException;
import org.openbase.jul.exception.InvalidStateException;
import org.openbase.jul.exception.NotAvailableException;
import org.openbase.jul.visual.swing.engine.draw2d.AbstractResourcePanel;
import org.openbase.planetsudo.geometry.Point2D;
import org.slf4j.LoggerFactory;

/**
 *
 * @author <a href="mailto:divine@openbase.org">Divine Threepwood</a>
 */
public class Mothership extends AbstractLevelObject implements ActionListener, MothershipInterface {

	public final static String MOTHERSHIP_FUEL_STATE_CHANGE = "FuelStateChange";
	public final static String MOTHERSHIP_SHIELD_STATE_CHANGE = "ShieldStateChange";
	public final static int MOTHERSHIP_FUEL_VOLUME = 15000;
	public final static int TOWER_FUEL_VOLUME = 2000;
	public final static int AGENT_FUEL_VOLUME = 12000;
	public final static int COMMANDER_BONUS_FUEL_VOLUME = 1000;
	public final static int MAX_AGENT_COUNT = 10; // range 0-9999
	public final static int BURNING_MOTHERSHIP = 20;
	public final static int BURNING_TOWER = 20;
    
    private final Logger logger = LoggerFactory.getLogger(getClass());
    
	public final Object AGENTLOCK = new Object();
	public final Object SUPPORT_CHANNEL_LOCK = new Object();
	private final Team team;
	private int fuel;
	private int strategyAgentCount;
	private int shield;
	private final Timer timer;
	private final Map<Integer, Agent> agents;
	private final List<Agent> supportChannel;
    private final Tower tower;
	private final TeamMarker teamMarker;
	private int mineCounter;

	public Mothership(final int id, final Team team, final AbstractLevel level) {
		super(id, team.getName() + Mothership.class.getSimpleName(), AbstractResourcePanel.ObjectType.Static, level, level.getMothershipBase(id).getPoint(), 100, 100, ObjectShape.Rec);
		logger.info("Create " + this);
		this.team = team;
		this.team.setMothership(this);
		this.agents = new HashMap<Integer, Agent>();
		this.supportChannel = new ArrayList<Agent>();
		this.teamMarker = new TeamMarker(team, level);
		this.timer = new Timer(50, this);
		this.mineCounter = 0;
        this.tower = new Tower(id, level, this);
	}

	@ Override
	public final void reset() {
		GUIController.setEvent(new PropertyChangeEvent(this, GUIController.LOADING_STATE_CHANGE, 1, "Lade " + team.getName() + " Mutterschiff"));
		fuel = MOTHERSHIP_FUEL_VOLUME;
		mineCounter = level.getMineCounter();

		final List<Agent> tmpCollection = new LinkedList<Agent>();
		synchronized (AGENTLOCK) {
			tmpCollection.addAll(agents.values());
		}
		for (Agent agent : tmpCollection) {
			agent.kill();
		}
		strategyAgentCount = team.getAgentCount();
		loadAgents();
		this.shield = 100;
	}

	public synchronized boolean orderMine() {
		if (mineCounter > 0) {
			mineCounter--;
			return true;
		}
		return false;
	}

	private void loadAgents() {
		GUIController.setEvent(new PropertyChangeEvent(this, GUIController.LOADING_STATE_CHANGE, strategyAgentCount, "Lade " + team.getName() + " Agent"));
		Agent agent;
		final int agentFuelVolume = (AGENT_FUEL_VOLUME / strategyAgentCount);
		int currentFuelVolume;
		boolean commanderFlag = true;
		for (int i = 0; i < strategyAgentCount; i++) {
			GUIController.setEvent(new PropertyChangeEvent(this, GUIController.LOADING_STEP, null, i));
			currentFuelVolume = commanderFlag ? agentFuelVolume + COMMANDER_BONUS_FUEL_VOLUME : agentFuelVolume;
			agent = new Agent(team.getName() + "Agent", commanderFlag, currentFuelVolume, this);
			commanderFlag = false;
			Agent replacedAgent;
			synchronized (AGENTLOCK) {
				replacedAgent = agents.put(agent.getId(), agent);
			}
			if (replacedAgent != null) {
				logger.error("Agent with id " + id + " already defined! Kill old instance.");
				replacedAgent.kill();
			}
		}
		synchronized (AGENTLOCK) {
			agentCount = agents.size();
			agentKeyArray = new Integer[agents.size()];
			agentKeyArray = agents.keySet().toArray(agentKeyArray);
		}
	}

	public int orderFuel(int fuel, final Agent agent) {
		if (agent == null || getBounds().contains(agent.getBounds())) {
			try {
				final int oldFuel = this.fuel;
				if (fuel <= 0) { // fuel emty
					fuel = 0;
				} else if (this.fuel < fuel) { // use last fuel
					fuel = this.fuel;
					this.fuel = 0;
					synchronized (this) {
						notifyAll();
					}
				} else {
					this.fuel -= fuel;
				}
				changes.firePropertyChange(MOTHERSHIP_FUEL_STATE_CHANGE, oldFuel, this.fuel);
			} catch (Exception ex) {
				logger.error("Could not order fuel!", ex);
			}
		} else {
			return 0;
		}
		return fuel;
	}

    @Override
	public int getFuel() {
		return fuel;
	}
    
    @Override
    public int getFuelVolume() {
        return TOWER_FUEL_VOLUME;
    }
    
    @Override
    public int getFuelInPercent() {
        return (fuel * 100) / TOWER_FUEL_VOLUME;
    }


    @Override
	public boolean hasFuel() {
		return fuel > 0;
	}

	protected void spendFuel(final int value) {
		if (value + fuel > MOTHERSHIP_FUEL_VOLUME) {
			fuel = MOTHERSHIP_FUEL_VOLUME;
		} else {
			fuel += value;
		}
		changes.firePropertyChange(MOTHERSHIP_FUEL_STATE_CHANGE, null, this.fuel);
	}
	public static final Object TILL_END_WAITER = new Object();

	public void waitTillGameEnd() {
		synchronized (this) {
			if (fuel == 0) {
				return;
			}
			try {
				synchronized (TILL_END_WAITER) {
					TILL_END_WAITER.wait();
				}
			} catch (InterruptedException ex) {
				logger.error("", ex);
			}
		}
	}

	public void startGame() {
		synchronized (AGENTLOCK) {
			for (Agent agent : new ArrayList<>(agents.values())) {
				agent.startGame();
			}
		}
	}
	private int agentIndex = 0;
	private Integer[] agentKeyArray;
	private Agent nextAgent;
	private int agentCount = 0;

	public void addActionPoint() {
		if (agentCount != 0) {
			agentIndex = (agentIndex + 1) % (agentCount);
			synchronized (AGENTLOCK) {
				nextAgent = agents.get(agentKeyArray[agentIndex]);
			}
			if (nextAgent != null) {
				nextAgent.addActionPoint();
			}
		}
	}

	public Team getTeam() {
		return team;
	}

	public int registerAgent() {
		synchronized (AGENTLOCK) {
			int i;
			for (i = 0; i <= agents.size(); i++) {
				if (!agents.containsKey(getId() * 10000 + i)) {
					break;
				}
			}
			if (i >= strategyAgentCount || i > 1000) {
				logger.error("Already to many agents alive.");
				return -1;
			}
			return getId() * 10000 + i;
		}
	}

	/**
	 * Method just for visual purpose
	 *
	 * @return
	 */
	public TeamMarker getTeamMarker() {
		return teamMarker;
	}

	@Override
	public int getAgentCount() {
		synchronized (AGENTLOCK) {
			return agents.size();
		}
	}

	protected void removeAgent(final Agent agent) {
		synchronized (AGENTLOCK) {
			agents.remove(agent.getId());
		}
	}

	public Point2D getAgentHomePosition() {
		return position.clone();
	}

	protected void passResource(final Agent agent) throws InvalidStateException {
		final Resource resource = agent.getResource();
		if (resource != null){
			if(getBounds().contains(agent.getBounds())) {
				team.addPoints(resource.use(agent));
			} else {
				resource.release();
			}
			
		}
	}

	public Collection<Agent> getAgents() {
		synchronized (AGENTLOCK) {
			return agents.values();
		}
	}

	public synchronized void attack() {
		logger.debug("Attack Mothership");
		if (shield > 0) {
			shield--;
			if (shield <= BURNING_MOTHERSHIP) {
				if (!timer.isRunning()) {
					timer.start();
					GameSound.MothershipExplosion.play();
				}
			}
			changes.firePropertyChange(MOTHERSHIP_SHIELD_STATE_CHANGE, null, shield);
		}
	}

	public synchronized void repair() {
		if (shield < 100) {
			shield++;
			if (shield > BURNING_MOTHERSHIP && timer.isRunning()) {
				timer.stop();
			}
			changes.firePropertyChange(MOTHERSHIP_SHIELD_STATE_CHANGE, null, shield);
		}
	}

	public boolean isBurning() {
		return shield < BURNING_MOTHERSHIP && hasFuel();
	}

	public int getShieldForce() {
		return shield;
	}

	public int getShieldPoints() {
		return shield / 2;
	}

	public boolean isMaxDamaged() {
		return shield == 0;
	}

	public boolean isDamaged() {
		return shield < 100;
	}
    
    public Tower getTower() {
        return tower;
    }

	@Override
	public void actionPerformed(final ActionEvent ex) {
		if (!GameManager.getInstance().isPause()) {
			orderFuel(Math.max(0, BURNING_MOTHERSHIP - shield), null);
		}
	}

	public int getAgentsAtHomePoints() {
		return ((getAgentsAtHomePosition() * 100) / agentCount);
	}

	public int getAgentsAtHomePosition() {
		int counter = 0;
		synchronized (AGENTLOCK) {
			for (Agent agent : agents.values()) {
				if (getBounds().contains(agent.getBounds()) || getBounds().intersects(agent.getBounds())) {
					counter++;
				}
			}
		}
		return counter;
	}

	public void callForSupport(final Agent agent) {
		synchronized (SUPPORT_CHANNEL_LOCK) {
			if (supportChannel.contains(agent)) {
				return;
			}
			supportChannel.add(agent);
			agent.setNeedSupport(true);
		}
	}

	public boolean needSomeoneSupport() {
		synchronized (SUPPORT_CHANNEL_LOCK) {
			return supportChannel.size() > 0;
		}
	}

	public Agent getAgentToSupport(final Agent helper) throws CouldNotPerformException {
		synchronized (SUPPORT_CHANNEL_LOCK) {
			if (supportChannel.isEmpty()) {
				throw new CouldNotPerformException("No support necessary.");
			}
			Agent supportCaller = null;
			int distance = Integer.MAX_VALUE;
			int tmpDistance;
			for (Agent a : supportChannel) {
				if (a == helper) { // do not help yourself
					continue;
				}
				tmpDistance = helper.getLevelView().getDistance(a);
				if (supportCaller == null || tmpDistance < distance) {
					supportCaller = a;
					distance = tmpDistance;
				}
			}

			if (supportCaller == null) {
				throw new CouldNotPerformException("No support possible.");
			}

			// remove caller from support channel if support possible
			if (supportCaller.getBounds().intersects(helper.getViewBounds())) {
				cancelSupport(supportCaller);
			}
			supportCaller.getLevelView().updateObjectMovement(true);
			return supportCaller;
		}
	}

	public void cancelSupport(final Agent agent) {
		synchronized (SUPPORT_CHANNEL_LOCK) {
			supportChannel.remove(agent);
			agent.setNeedSupport(false);
		}
	}

	@Override
	public boolean isMarkerDeployed() {
		return teamMarker.isPlaced();
	}

	@Override
	public void clearMarker() {
		teamMarker.clear();
	}

	protected void placeMarker(final Point2D position) {
		teamMarker.place(position);
	}

	protected TeamMarker getMarker() throws CouldNotPerformException {
		return teamMarker.getMarker();
	}

	public void setGameOverSoon() {
		for(Agent agent : agents.values()) {
			agent.setGameOverSoon();
		}
	}
}

File: src/main/kotlin/org/openbase/planetsudo/level/levelobjects/Mothership.kt
package org.openbase.planetsudo.level.levelobjects

import org.openbase.jul.exception.CouldNotPerformException
import org.openbase.jul.exception.InvalidStateException
import org.openbase.jul.visual.swing.engine.draw2d.AbstractResourcePanel
import org.openbase.planetsudo.game.GameManager
import org.openbase.planetsudo.game.GameSound
import org.openbase.planetsudo.game.Team
import org.openbase.planetsudo.geometry.Base2D
import org.openbase.planetsudo.geometry.Point2D
import org.openbase.planetsudo.level.AbstractLevel
import org.openbase.planetsudo.main.GUIController
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.beans.PropertyChangeEvent
import java.util.*
import java.util.concurrent.locks.ReentrantReadWriteLock
import javax.swing.Timer
import kotlin.concurrent.read
import kotlin.concurrent.write
import kotlin.math.max

/**
 *
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */
class Mothership(id: Int, team: Team, level: AbstractLevel) :
    AbstractLevelObject(
        id,
        team.name + Mothership::class.java.simpleName,
        AbstractResourcePanel.ObjectType.Static,
        level,
        level.getMothershipBase(id).point,
        100.0,
        100.0,
        ObjectShape.Rec,
    ),
    ActionListener,
    MothershipInterface {
    private val logger: Logger = LoggerFactory.getLogger(javaClass)

    private val agentLock = ReentrantReadWriteLock()
    private val supportChannelLock = ReentrantReadWriteLock()
    val base: Base2D = level.getMothershipBase(id)

    @JvmField
    val team: Team

    override val fuelVolume: Int
        get() = MOTHERSHIP_FUEL_VOLUME
    override var fuel: Int = 0
        private set
    private var strategyAgentCount = 0
    override var shieldForce: Int = 0
        private set
    private val timer: Timer
    var agents: List<Agent>
        private set
    private val supportChannel: MutableList<Agent>

    @JvmField
    val tower: Tower

    /**
     * Method just for visual purpose
     *
     * @return
     */
    @JvmField
    val teamMarker: TeamMarker
    private var mineCounter: Int

    public override fun reset() {
        GUIController.setEvent(
            PropertyChangeEvent(
                this,
                GUIController.LOADING_STATE_CHANGE,
                1,
                "Lade " + team.name + " Mutterschiff",
            ),
        )
        fuel = MOTHERSHIP_FUEL_VOLUME
        mineCounter = level.mineCounter
        agentLock.write {
            agents.forEach { it.kill() }
            strategyAgentCount = team.agentCount
            loadAgents()
        }
        this.shieldForce = 100
    }

    @Synchronized
    fun orderMine(): Boolean {
        if (mineCounter > 0) {
            mineCounter--
            return true
        }
        return false
    }

    private fun loadAgents() {
        agentLock.write {
            GUIController.setEvent(
                PropertyChangeEvent(
                    this,
                    GUIController.LOADING_STATE_CHANGE,
                    strategyAgentCount,
                    "Lade " + team.name + " Agent",
                ),
            )
            var agent: Agent
            val agentFuelVolume = (AGENT_FUEL_VOLUME / strategyAgentCount)
            var currentFuelVolume: Int
            var commanderFlag = true
            for (i in 0 until strategyAgentCount) {
                GUIController.setEvent(PropertyChangeEvent(this, GUIController.LOADING_STEP, null, i))
                currentFuelVolume =
                    if (commanderFlag) agentFuelVolume + COMMANDER_BONUS_FUEL_VOLUME else agentFuelVolume
                agent = Agent(team.name + "Agent", commanderFlag, currentFuelVolume, this)
                commanderFlag = false
                agents = agents.plus(agent)
            }
        }
    }

    fun orderFuel(fuel: Int, agent: Agent?): Int {
        var fuel = fuel
        if (agent == null || bounds.contains(agent.bounds)) {
            try {
                val oldFuel = this.fuel
                if (fuel <= 0) { // fuel emty
                    fuel = 0
                } else if (this.fuel < fuel) { // use last fuel
                    fuel = this.fuel
                    this.fuel = 0
                    synchronized(this) {
                        (this as Object).notifyAll()
                    }
                } else {
                    this.fuel -= fuel
                }
                changes.firePropertyChange(MOTHERSHIP_FUEL_STATE_CHANGE, oldFuel, this.fuel)
            } catch (ex: CouldNotPerformException) {
                logger.error("Could not order fuel!", ex)
            }
        } else {
            return 0
        }
        return fuel
    }

    override val fuelInPercent: Int
        get() = (fuel * 100) / fuelVolume

    override fun hasFuel(): Boolean {
        return fuel > 0
    }

    fun spendFuel(value: Int) {
        if (value + fuel > MOTHERSHIP_FUEL_VOLUME) {
            fuel = MOTHERSHIP_FUEL_VOLUME
        } else {
            fuel += value
        }
        changes.firePropertyChange(MOTHERSHIP_FUEL_STATE_CHANGE, null, this.fuel)
    }

    fun waitTillGameEnd() {
        synchronized(this) {
            if (fuel == 0) {
                return
            }
            synchronized(TILL_END_WAITER) {
                (TILL_END_WAITER as Object).wait()
            }
        }
    }

    fun startGame() {
        agentLock.read {
            agents.forEach { it.startGame() }
        }
    }

    private var agentIndex = 0

    //    private var agentKeyArray: Array<Int>
    private var nextAgent: Agent? = null

    init {
        logger.info("Create $this")
        this.team = team
        this.team.mothership = this
        this.agents = emptyList()
        this.supportChannel = ArrayList()
        this.teamMarker = TeamMarker(team, level)
        this.timer = Timer(50, this)
        this.mineCounter = 0
        this.tower = Tower(id, level, this)
    }

    fun addActionPoint() {
        agentLock.read {
            if (agents.isNotEmpty()) {
                agentIndex = (agentIndex + 1) % (agents.size)
                nextAgent = agents[agentIndex]
            }
            nextAgent?.addActionPoint()
        }
    }

    fun computeNextAgentId(): Int = agentLock.read { id * 10000 + agents.size }

    override val agentCount: Int get() = agentLock.read { agents.size }

    fun removeAgent(agent: Agent) = agentLock.write {
        agents
            .firstOrNull { it.id == agent.id }
            ?.also { agents = agents.minus(it) }
    }

    val agentHomePosition: Point2D
        get() = position.clone()

    @Throws(InvalidStateException::class)
    fun passResource(agent: Agent) {
        val resource = agent.getResource()
        if (resource != null) {
            if (bounds.contains(agent.bounds)) {
                team.addPoints(resource.use(agent))
            } else {
                resource.release()
            }
        }
    }

    @Synchronized
    fun attack() {
        logger.debug("Attack Mothership")
        if (shieldForce > 0) {
            if (shieldForce == 100) {
                GameSound.MothershipUnderAttack.play()
            }

            shieldForce--
            if (shieldForce <= BURNING_MOTHERSHIP) {
                if (!timer.isRunning) {
                    timer.start()
                    GameSound.MothershipExplosion.play()
                }
            }
            changes.firePropertyChange(MOTHERSHIP_SHIELD_STATE_CHANGE, null, shieldForce)
        }
    }

    @Synchronized
    fun repair() {
        if (shieldForce < 100) {
            shieldForce++
            if (shieldForce > BURNING_MOTHERSHIP && timer.isRunning) {
                timer.stop()
            }
            changes.firePropertyChange(MOTHERSHIP_SHIELD_STATE_CHANGE, null, shieldForce)
        }
    }

    override val isBurning: Boolean
        get() = shieldForce < BURNING_MOTHERSHIP && hasFuel()

    val shieldPoints: Int
        get() = shieldForce / 2

    override val isMaxDamaged: Boolean
        get() = shieldForce == 0

    override val isDamaged: Boolean
        get() = shieldForce < 100

    override fun actionPerformed(ex: ActionEvent) {
        if (!GameManager.gameManager.isPause) {
            orderFuel(
                max(0.0, (BURNING_MOTHERSHIP - shieldForce).toDouble()).toInt(),
                null,
            )
        }
    }

    val agentsAtHomePoints: Int
        get() = ((agentsAtHomePosition * 100) / strategyAgentCount)

    private val agentsAtHomePosition: Int
        get() {
            agentLock.read {
                var counter = 0
                agents.forEach { agent ->
                    if (bounds.contains(agent.bounds) || bounds.intersects(agent.bounds)) {
                        counter++
                    }
                }
                return counter
            }
        }

    fun callForSupport(agent: Agent) {
        supportChannelLock.write {
            if (supportChannel.contains(agent)) {
                return
            }
            supportChannel.add(agent)
            agent.setNeedSupport(true)
        }
    }

    override fun needSomeoneSupport(): Boolean {
        supportChannelLock.read {
            return supportChannel.size > 0
        }
    }

    @Throws(CouldNotPerformException::class)
    fun getAgentToSupport(helper: Agent): Agent {
        supportChannelLock.read {
            if (supportChannel.isEmpty()) {
                throw CouldNotPerformException("No support necessary.")
            }

            val supportCaller: Agent = supportChannel
                .filterNot { it === helper } // do not help yourself
                .minByOrNull { helper.levelView?.getDistance(it) ?: Int.MAX_VALUE }
                ?: throw CouldNotPerformException("No support possible.")

            // remove caller from support channel if support is possible
            if (supportCaller.bounds.intersects(helper.viewBounds)) {
                cancelSupport(supportCaller)
            }
            supportCaller.levelView?.updateObjectMovement()
            return supportCaller
        }
    }

    fun cancelSupport(agent: Agent) {
        supportChannelLock.write {
            supportChannel.remove(agent)
            agent.setNeedSupport(false)
        }
    }

    override val isMarkerDeployed: Boolean
        get() = teamMarker.isPlaced

    override fun clearMarker() {
        teamMarker.clear()
    }

    fun placeMarker(position: Point2D) {
        teamMarker.place(position)
    }

    @get:Throws(CouldNotPerformException::class)
    val marker: TeamMarker
        get() = teamMarker.marker

    fun setGameOverSoon() = agentLock.read { agents.forEach { it.setGameOverSoon() } }

    companion object {
        const val MOTHERSHIP_FUEL_STATE_CHANGE: String = "FuelStateChange"
        const val MOTHERSHIP_SHIELD_STATE_CHANGE: String = "ShieldStateChange"
        const val MOTHERSHIP_FUEL_VOLUME: Int = 15000
        const val AGENT_FUEL_VOLUME: Int = 12000
        const val COMMANDER_BONUS_FUEL_VOLUME: Int = 1000
        const val MAX_AGENT_COUNT: Int = 10 // range 0-9999
        const val BURNING_MOTHERSHIP: Int = 20
        const val BURNING_TOWER: Int = 20

        val TILL_END_WAITER: Any = Any()
    }
}
